### PR TITLE
ISLE: Allow shadowing in let expressions

### PR DESCRIPTION
**Pull request by elliottt**
Currently, a let expression in ISLE currently requires unique names for each declaration. If you are sequencing a number of effects together using a let expression, you'll end up with multiple declarations whose results you don't need. Here's an example of this from the s390x backend: https://github.com/bytecodealliance/wasmtime/blob/8e9e9c52a1da9c661ad8a25210e38826bf8f7362/cranelift/codegen/src/isa/s390x/inst.isle#L3057-L3064

This PR allows shadowing in let expressions, enabling the previous example to be refactored in the following way:
```
;; Use a boolean condition to select between two registers.
(decl select_bool_reg (Type ProducesBool Reg Reg) Reg)
(rule (select_bool_reg ty (ProducesBool.ProducesBool producer cond) reg_true reg_false)
      (let ((dst WritableReg (temp_writable_reg ty))
            (_ Unit (emit_producer producer))
            (_ Unit (emit_mov ty dst reg_false))
            (_ Unit (emit_consumer (emit_cmov_reg ty dst cond reg_true))))
        dst))
```

### CODE REVIEW

**Comment by elliottt** [comment]
It might be better to implement full shadowing instead, which would resolve #3526.

**Review by cfallin — APPROVED** [review]
This looks reasonable to me and I'm happy to see it landed as-is if you'd like; or, a full implementation of binding shadowing is fine too as it would subsume this. (The tricky bit with the latter will be the "pop" at the end of the nested def I suspect; either a scoped hashmap or a "save the old binding and restore it" bit of logic in the sema lowering for `let` will be necessary, I think.)

**Comment by elliottt** [comment]
@cfallin I went the route of implementing shadowing instead, as it seems like a more complete solution to the problem. To resolve the issue of restoring the previous bindings, I turned the `TypeEnv::sym_map` field into a vector of maps, ensuring that the vector is never empty. I'd be happy to replace the vector of maps with a more efficient implementation :+1:

File: cranelift/isle/isle/isle_examples/pass/let_shadowing.isle [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+(decl test3 (u64) u64)
+(rule (test3 x)
+      (let ((test3 u64 x))
+        test3))
+
+;; Shadowing another let-bound name
+(decl test4 (u64) u64)
+(rule (test4 x)
+      (let ((val u64 x)
+            (val u64 23))
```
Comment (cfallin) [inline cranelift/isle/isle/isle_examples/pass/let_shadowing.isle:29]: Would you mind turning at least this case into a run-test and invoking the rule from a test toplevel, asserting that the most recent binding wins?

**Review by cfallin — CHANGES_REQUESTED** [review]
Thanks for pushing further and doing full shadowing!

However I think there may be a simpler way to do this. I didn't expect the symbol-interning itself to become aware of nested scopes, and in fact that could (?) lead to unexpected behavior later if we remove a binding. Basically I had meant for that to be an interning layer so that we could use `Sym` indices everywhere as a substitute for string identifiers.

Instead I had expected `bindings` to handle the scope entry and exit. It was actually structured originally with the intent of supporting shadowing some day: it remains an ordered list of bindings (in order of introduction). We truncate away all bindings within the `let` after we process the `let`'s body. I suspect all we need to do is to remove the "am I shadowing another binding" check (line 1954 in the original `sema.rs`), and ensure that the search for a binding occurs as a reverse search from the end (I *believe* it already might). Does that make sense?

**Comment by elliottt** [comment]
> Thanks for pushing further and doing full shadowing!
> 
> However I think there may be a simpler way to do this. I didn't expect the symbol-interning itself to become aware of nested scopes, and in fact that could (?) lead to unexpected behavior later if we remove a binding. Basically I had meant for that to be an interning layer so that we could use `Sym` indices everywhere as a substitute for string identifiers.
> 
> Instead I had expected `bindings` to handle the scope entry and exit. It was actually structured originally with the intent of supporting shadowing some day: it remains an ordered list of bindings (in order of introduction). We truncate away all bindings within the `let` after we process the `let`'s body. I suspect all we need to do is to remove the "am I shadowing another binding" check (line 1954 in the original `sema.rs`), and ensure that the search for a binding occurs as a reverse search from the end (I _believe_ it already might). Does that make sense?

This makes sense, and is indeed a much simpler approach. The var lookup already searches backwards through the bindings, so this really might be as simple as removing the check for duplicate bindings.

File: cranelift/isle/isle/isle_examples/pass/let_shadowing.isle (hunk 1 above)
Comment (elliottt) [reply cranelift/isle/isle/isle_examples/pass/let_shadowing.isle]: I reworked it as a `link` test in the `isle/isle_examples/link` directory, and modified the test runner to also execute the output of `rustc` on these tests.

Comment (cfallin) [reply cranelift/isle/isle/isle_examples/pass/let_shadowing.isle]: Ah, I think you may need to rebase -- as of Friday there is now a notion of a run-test already and a separate `isle_examples/run` directory :-)

File: cranelift/isle/isle/src/sema.rs [hunk 2]
```diff
@@ -1951,9 +1951,6 @@ impl TermEnv {
                 for def in defs {
                     // Check that the given variable name does not already exist.
                     let name = tyenv.intern_mut(&def.var);
-                    if bindings.vars.iter().any(|bv| bv.name == name) {
-                        tyenv.report_error(pos, format!("Variable '{}' already bound", def.var.0));
-                    }
```
Comment (elliottt) [inline cranelift/isle/isle/src/sema.rs:1956]: It's hilarious that this is what the PR boils down to :)

File: cranelift/isle/isle/isle_examples/pass/let_shadowing.isle (hunk 1 above)
Comment (elliottt) [reply cranelift/isle/isle/isle_examples/pass/let_shadowing.isle]: Nice! I rebased and removed the changes I introduced to run the `link` tests :+1:

File: cranelift/isle/isle/src/sema.rs (hunk 2 above)
Comment (cfallin) [reply cranelift/isle/isle/src/sema.rs:1956]: Something about journeys vs. destinations etc etc ... but, yes, I'm happy it was a simple change!

**Review by cfallin — APPROVED** [review]
LGTM, thanks!

**Pull request merged**